Update to the latest version of the crypto refresh

lubux opened the pull request:
Includes the updates for the latest merged pull requests in the crypto refresh draft:

- PKESK [!304](https://gitlab.com/openpgp-wg/rfc4880bis/-/merge_requests/304) 
- OPS [!305](https://gitlab.com/openpgp-wg/rfc4880bis/-/merge_requests/305) 
- Cleartext [!313](https://gitlab.com/openpgp-wg/rfc4880bis/-/merge_requests/313) 

Further changes:
- The PKESK constructor now sets the version to null
- `createSignaturePackets` has an additional optional argument `signatureSalts` that allows to pass previously generated salts. This argument is used for signature creation with ops packets.
- `createSignaturePacket` has an additional optional argument `salt` that allows to pass a previously generated salt.

twiss committed:
For v6 keys, create direct-key signature for key properties

Store key flags, features and preferences in a direct-key signature
instead of user ID signatures, for V6 keys.

twiss committed:
Don't require User IDs for v6 keys

twiss committed:
Generate SEIPD v2 flag instead of AEAD flag

The AEAD Encrypted Data packet has been removed from the draft
in favor of version 2 of the Sym. Encrypted Integrity Protected
Data packet. It also has a new feature flag to match.

twiss committed:
Don't generate v5 keys flag

This flag has been removed from the draft specification.

twiss committed:
Add (non-experimental) GCM

Also, set it as the preferred AEAD algorithm.

twiss committed:
Add preferred ciphersuites subpacket

This subpacket replaces both symmetric algorithm preferences and
AEAD algorithm preferences when AEAD is supported, by providing
sets of preferred symmetric and AEAD algorithm pairs.

We still keep the symmetric algorithm preferences in case AEAD is
not supported.

twiss committed:
Remove embedded Web Streams ponyfill

Require the application to load a polyfill instead.

twiss committed:
Update ESLint globals syntax

twiss committed:
Use globalThis for setting test helper functions globals

twiss committed:
Don't generate armor checksum lines

twiss committed:
Remove compression algorithms from preferences

twiss committed:
Remove AES-192 from preferred symmetric algorithms

Chrome's Web Crypto implementation doesn't support it, and it
seems unnecessary to list it when AES-256 is available.

twiss committed:
For v6 keys, check direct-key signature for key properties

Key flags, expiration time, algorithm preferences, et cetera, are now
read from the direct-key signature instead of the primary User ID
binding signature for v6 keys.

This also requires a direct-key signature to be present for v6 keys.

twiss committed:
Tests: update expected algorithm preferences

twiss committed:
Update fallback (mandatory) AEAD algorithm to OCB

This has been changed in the crypto refresh.

twiss committed:
Remove armor checksum check

The crypto refresh says that we MUST NOT reject messages where the
CRC24 checksum is incorrect. So, we remove the check for it.

Also, remove the checksumRequired config.

twiss committed:
Implement SEIPD v2

twiss committed:
Implement Padding Packet

twiss committed:
Support SKESK v6 as per the latest crypto refresh

The latest crypto refresh specifies an HKDF step to be used for
deriving the key to encrypt the session key with.

It also specifies two additional length fields.

twiss committed:
Add support for v6 key packets

Compared to v5 keys, v6 keys contain additional length fields to aid in
parsing the key, but omit the secret key material length field.

Additionally, unencrypted v6 secret key packets don't include the count
of the optional fields, as per the updated crypto refresh. Since they
are always absent, the count is not needed.

twiss committed:
Add support for v6 signatures

Compared to v5 signatures, v6 signatures include a salt, and the
subpacket lengths are increased from 2 to 4 bytes.

twiss committed:
Replace config.v5Keys with config.v6Keys flag

twiss committed:
Look up preferred ciphersuite in one go

Instead of calling getPreferredAlgo('symmetric') and
getPreferredAlgo('aead'), we define and call getPreferredCipherSuite()
to determine the preferred symmetric and AEAD algorithm.

Additionally, we remove isAEADSupported(), instead we return
aeadAlgorithm = 0 from getPreferredCipherSuite() if AEAD is not
supported.

And finally, we define getPreferredCompressionAlgo() to replace
getPreferredAlgo('compression').

twiss committed:
Add support for PKESK v6

twiss committed:
When decrypting a v6 PKESK in constant-time, use the v2 SEIPD algorithm

Rather than using the config to determine which algorithms to try
to decrypt session keys for, try the algorithm we know the message
was encrypted with.

twiss committed:
Move the PKCS#5 padding into the algorithms

Rather than padding in the PKESK packet, add padding inside the
public-key algorithms, so that we can later add algorithms without
any padding.

Additionally, simplify the constant-time decryption logic, by always
using the assumed algorithm, rather than doing a constant-time select
based on whether the decrypted algorithm identifier was correct.

twiss committed:
Remove checksum in unencrypted v6 secret keys

twiss committed:
Update initial hash octet for v6 fingerprints

twiss committed:
Update web-stream-tools

larabr committed:
CI: temporarily enable for PRs to v6 branch

larabr committed:
`crypto-refresh`: add support for Argon2 S2K (#1597)

In terms of API, this feature is backwards compatible, no breaking changes.
However, since a Wasm module is loaded for the Argon2 computation, browser apps
might need to make changes to their CSP policy in order to use the feature.

Newly introduced config fields:
- `config.s2kType` (defaulting to `enums.s2k.iterated`): s2k to use on
password-based encryption as well as private key encryption;
- `config.s2kArgon2Params` (defaulting to "uniformly safe settings" from Argon
RFC): parameters to use on encryption when `config.s2kType` is set to
`enums.s2k.argon2`;

larabr committed:
CI: drop Node 14, add Node 20 (canary)

larabr committed:
Throw intelligible error on GCM authentication failure, fix/refactor test for modification detection on decryption

larabr committed:
Fix floating promise rejection in AEAD message decryption

larabr committed:
Ignore race condition in AEAD message decryption tests

larabr committed:
CI: don't test on Node 20

Need to update Sinon, plus need to investigate AEAD decryption errors

larabr committed:
Fix race condition in AEAD message decryption

twiss committed:
Appease linter

twiss committed:
Return object from getPreferredCipherSuite instead of array

Rather than returning [symmetricAlgo, aeadAlgo] from
getPreferredCipherSuite, return { symmetricAlgo, aeadAlgo }.

Also, return aeadAlgo: undefined rather than 0 to indicate no AEAD
algorithm is used (CFB is used instead).

larabr committed:
Support AEAD encryption for v4 and v6 private keys

lubux committed:
Add support for v6 one-pass signature packets

Introduces v6 one-pass signature packets required for v6 signatures.
Includes the changes from !305 of the crypto refresh:
https://gitlab.com/openpgp-wg/rfc4880bis/-/merge_requests/305

lubux committed:
Update PKESK v6 to the latest version of the crypto refresh

Implement the changes from !304 of the crypto refresh:
https://gitlab.com/openpgp-wg/rfc4880bis/-/merge_requests/304

"Introduces an explicit length octet for the fingerprint.
This length also covers the key version number.  This has the
advantage of nicely covering the anonymous recipient case, where a
zero length indicates not having a fingerprint in the packet without
having to resort to a fictitious key version zero."

lubux committed:
Remove cleartext message header check

The latest version of the crypto refresh (i.e., !313, !314) specifies that
the  "Hash" header is depricated. An implementation that is verifying
a cleartext signed message MUST ignore this header.
Thus, the header check can be removed.
This commit does not change the writing part of cleartext messages.

larabr commented on src/packet/public_key_encrypted_session_key.js:
```diff
@@ -108,8 +115,13 @@ class PublicKeyEncryptedSessionKeyPacket {
     if (this.version === 6) {
-      arr.push(new Uint8Array([this.publicKeyVersion]));
-      arr.push(this.publicKeyFingerprint);
+      if (this.publicKeyFingerprint === null) {
```
```suggestion
      if (this.publicKeyFingerprint !== null) {
```
?

lubux replied:
Yes correct, thanks :)

larabr commented on src/packet/one_pass_signature.js:
```diff
@@ -40,8 +38,8 @@ class OnePassSignaturePacket {
-    /** A one-octet version number.  The current version is 3. */
-    this.version = null;
+    /** A one-octet version number.  The current versions are 3 and 6. */
+    this.version = 3;
```
We're moving away from initialising values in the constructor that are later overwritten (either by `read` or when signing with a v6 key in this case), so I'd leave this null (same as with signature packets)

lubux replied:
Makes sense. I copied it from the PKESK packet implementation. I will change it there as well.

lubux replied:
Ok, it seems that this change breaks some tests. So, I will leave it at 3.

larabr replied:
The fact that it breaks some tests is a bad sign, since it was `null` to begin with 😓  there might be a missing overwrite based on key version 

Edit: actually, before `VERSION` was often used over `this.version`, so it might explain why it breaks stuff. Still I think it's safer to not initialise.

lubux replied:
The issue was that objects were initialised with the constructor without setting the version field. Fixed now.

larabr commented on src/packet/public_key_encrypted_session_key.js:
```diff
@@ -108,8 +115,13 @@ class PublicKeyEncryptedSessionKeyPacket {
-      arr.push(this.publicKeyFingerprint);
+      if (this.publicKeyFingerprint !== null) {
+        arr.push(new Uint8Array([this.publicKeyFingerprint.length + 1]));
+        arr.push(new Uint8Array([this.publicKeyVersion]));
```
```suggestion
        arr.push(new Uint8Array([
          this.publicKeyFingerprint.length + 1,
          this.publicKeyVersion
        ]));
```
(minor)

lubux committed:
Add decryption test for a gopenpgp message with v6 packets

lubux committed:
Add v6 decrypt/verify test produced by gopenpgp for openpgp.decrypt

larabr commented on test/general/packet.js:
```diff
@@ -600,6 +600,115 @@ module.exports = () => describe('Packet', function() {
     });
```
We should add a test that generates a v6 signed message and confirms that the OPS is the correct version and it is used to verify

lubux replied:
Added a test that checks if a one-pass signature packet is correctly generated when (inline) signing with a v6 key.

lubux commented on src/packet/one_pass_signature.js:
```diff
@@ -133,7 +170,8 @@ class OnePassSignaturePacket {
       correspondingSig.signatureType !== this.signatureType ||
       correspondingSig.hashAlgorithm !== this.hashAlgorithm ||
       correspondingSig.publicKeyAlgorithm !== this.publicKeyAlgorithm ||
-      !correspondingSig.issuerKeyID.equals(this.issuerKeyID)
```
TODO: Check that the versions match

lubux committed:
Set version in PKESK constructor to null

lubux committed:
Only emit Hash header below V6 for cleartext messages

The latest version of the crypto refresh (i.e., !313, !314) specifies that
the "Hash" header is depricated. This commit changes that the Hash header
is only generated if a cleartext message contains a non-V6 signature.

larabr commented on src/packet/public_key_encrypted_session_key.js:
```diff
@@ -67,13 +72,36 @@ class PublicKeyEncryptedSessionKeyPacket {
+      const sizeFields = bytes[offset++];
+      if (sizeFields !== 0) {
+        this.publicKeyVersion = bytes[offset++];
+        const fingerprintLength = this.publicKeyVersion ? (this.publicKeyVersion >= 5 ? 32 : 20) : 0;
```
```suggestion
        const fingerprintLength = this.publicKeyVersion >= 5 ? 32 : 20;
```
public key version of 0 has been removed in favour of the explicit size

larabr replied:
and we could actually use the `sizeFields - 1` value instead of hardcoding the fingerprint length

larabr commented on src/packet/public_key_encrypted_session_key.js:
```diff
@@ -83,11 +111,27 @@ class PublicKeyEncryptedSessionKeyPacket {
+    ];
+
+    if (this.version === 6) {
+      if (this.publicKeyFingerprint !== null) {
```
I think it might be easier to follow what's going on by referencing the wildcard keyID?
```suggestion
      if (!this.publicKeyID.isWildcard()) {
```

larabr commented on src/packet/signature.js:
```diff
@@ -195,6 +212,15 @@ class SignaturePacket {
 
+    if (this.version === 6) {
+      const saltLength = saltLengthForHash(this.hashAlgorithm);
+      if (this.salt === null) {
```
This conditional regeneration seems risky; to double check if needed.

larabr commented on src/cleartext.js:
```diff
@@ -159,44 +167,14 @@ export async function readCleartextMessage({ cleartextMessage, config, ...rest }
-    return true;
-  };
-
+function verifyHeaders(headers) {
```
We can now inline this

larabr commented:
I've cherry picked the changes to the target branch manually 👍 We'll also continue the review on the original PR since I cannot push to lubux's branch
